### ISSUE TITLE
redirect the stdout and stderr for each individual prep job to their own log.txt

### ISSUE THREAD

**Issue by apayne97** [choderalab/asapdiscovery#119]
# Description
I have now implemented using logging which outputs a log to individual log.txt files in each individual prepped protein directory. this is nice but it doesn't actually accomplish the potentially most useful thing, which is re-directing the stdout and stderr for each individual prep job to their own log.txt files. Right now when this is run in parallel on the cluster, all the error messages get overlapped in the log for the cluster script.

## Conclusion
The improvements suggested by @hmacdope in #171 work well, with the main trick being these two lines:
```
errfs = oechem.oeofstream(os.path.join(out_dir, f"openeye_{logname}-log.txt"))
oechem.OEThrow.SetOutputStream(errfs)
```
https://docs.eyesopen.com/toolkits/python/oechemtk/OESystemConstants/OEThrow.html#OESystem::OEThrow

# To-Do
- [ ] #171 
- [x] implement for mers prep script in #104 
- in the future this should be done this way for all future scripts that run things in parallel using openeye

## Related Issues / PRs
- [ ] #104

**Comment by apayne97** [comment]
@ijpulidos @mikemhenry @JenkeScheen do you have suggestions for the best way to do this?

**Comment by apayne97** [comment]
@hmacdope would you be able to help with this? this is for the `prep_proteins.py` script in the branch #104

**Comment by hmacdope** [comment]
Yep!

**Comment by JenkeScheen** [comment]
I'll leave the programmatic design choices to @hmacdope but I think it could be useful to write any logfiles that contain *hard* errors (i.e. ones where jobs have failed) to a more high-level location such as `./output/error_logs/` (possibly still writing them to protein folder files). I realise that may be a bit crowded but it's nice to not have to sift through all protein folders.

**Comment by apayne97** [comment]
@JenkeScheen it's pretty easy to `grep failed */*.txt` to see which ones failed! right now OpenEye warnings and errors are all getting written to the top level and its just incomprehensible. The only way in which printing to the top level would be useful would be if all error messages come with the job / complex / etc information in the line.

However, it would be pretty easy to have in the `prep_mp` part of the script just a prefix for all the output error messages that includes the complex id, and then write out to the top level. But I don't think that is any more helpful than grepping the directory?

**Comment by ijpulidos** [comment]
As far as I can tell `prep_mp` is logging to some `log.txt` and I guess when we run it in parallel this single file gets written by all the processes at the same time. If so, I wonder if it would be enough for the logging [file handler in `prep_mp`](https://github.com/choderalab/covid-moonshot-ml/blob/29f76821897702aa35d3be9b1d35137aa26e488d/asapdiscovery/docking/modeling.py#L711) to just include some information (maybe `ref_prot` or more) in the filename for the logger.

**Comment by apayne97** [comment]
The problem is that currently `logging` doesn't capture the OpenEye error output, just the stuff I'm explicitly writing out. The most useful thing for me would be to have it write out the OpenEye error out to separate files

**Comment by hmacdope** [comment]
@apayne97 I have added in a prefix to the 'log.txt' and `align_test.pdb` so that there is hopefully no multiprocessing issues as @ijpulidos suggested.  I pushed straight to the branch, hope that is ok. 

However I just wanted to confirm exactly what we want. Do you want to capture `stderr` and `stdout` for each multiprocessing job to its **own** logfile?  Or are we just trying to capture the output to some toplevel logfile in general.

@JenkeScheen I'm a little new to the workflow so bear with me. What indicates a failed job exactly?  Are there some files I could use to test locally?

**Comment by JenkeScheen** [comment]
Agreed with @apayne97, grepping should be fine for now.

**Comment by hmacdope** [comment]
I will close this for now then, feel free to re-open if still needs work.

**Comment by apayne97** [comment]
@hmacdope sorry for the delay on this!

> However I just wanted to confirm exactly what we want. Do you want to capture stderr and stdout for each multiprocessing job to its own logfile? Or are we just trying to capture the output to some toplevel logfile in general.

When running as a job on lilac, the output automatically gets written to the logfile in the batch script. I would like the logs to get written to the individual log files instead.

I have now added a test example in https://github.com/choderalab/asapdiscovery/pull/104/commits/8f322862fb0ca714f7427286f7d5402b4ed7aa33

Currently, print statements are getting written out to `asapdiscovery/modeling/tests/prep_mpro_test/prepped/Mpro-P2660_0A_EDG-MED-b1ef7fe3-1/Mpro-P2660_0A_EDG-MED-b1ef7fe3-1-log.txt`, but all the stuff like this:
```
DPI: 0.15, RFree: 0.24, Resolution: 1.64
Processing BU # 1 with title: , chains ABW
Found unresolved C-terminal with 1 residues after PHE 305   B 2  , with sequence Q
Found unresolved C-terminal with 1 residues after PHE 305   A 1  , with sequence Q
Warning: No loop database was provided
Warning: Failed.For LIG, could not update 'CNC(=O)CN1CC2(CCN(C2=O)c3cncc4c3cc(cc4)Cl)c5cc(ccc5C1=O)Cl' to match 'c1ccc-2c(c1)Cc3c2[nH]nc3c4ccncc4'
Warning: Failed.For LIG, could not update 'CNC(=O)CN1CC2(CCN(C2=O)c3cncc4c3cc(cc4)Cl)c5cc(ccc5C1=O)Cl' to match 'c1ccc-2c(c1)Cc3c2n[nH]c3c4ccncc4'
```
is not.

**Comment by hmacdope** [comment]
@apayne97 sorry for the delay, I'll hopefully tackle this today :)

**Comment by hmacdope** [comment]
Hmmmm Ive made **some** progress but also hit a few snags. I'm fairly sure that the part in question may be written on the C++ level by OE's SWIG bindings and not at the python level so It may be impossible to capture by python anyway as the logger isn't visible to the `LoggerManager`, ie its written to a different `ostream` (see here for some discussion https://stackoverflow.com/questions/24277488/in-python-how-to-capture-the-stdout-from-a-c-shared-library-to-a-variable) . I'm honestly not sure if its worth digging into the OE to fins out exactly how to get around this?

**Comment by hmacdope** [comment]
For future reference you can use this setup to capture stdout to a logger, but since that doesn't actually fix your issue @apayne97 I don't think I will PR. 


```python
import logging
import sys
import os


class StreamToLogger(object):
    """
    Fake file-like stream object that redirects writes to a logger instance.
    """
    def __init__(self, logger, level):
       self.logger = logger
       self.level = level
       self.linebuf = ''

    def write(self, buf):
       for line in buf.rstrip().splitlines():
          self.logger.log(self.level, line.rstrip())

    def flush(self):
        pass



class StreamRedirectedLogger:
    """
    A logger that captures stdout.
    """
    def __init__(self, name: str, logfile: str = None):
        self.name = name
        self.logfile = logfile
        if self.logfile is None:
            self.logfile = self.name + "-log.txt"
    
        self.formatter = logging.Formatter('%(asctime)s | %(levelname)s | %(message)s')
        self.logger = logging.getLogger(self.logfile)
        self.logger.setLevel(logging.DEBUG)

        self.fh = logging.FileHandler(self.logfile, mode='w')
        self.fh.setLevel(logging.DEBUG)
        self.fh.setFormatter(self.formatter)
        self.logger.addHandler(self.fh)
    
        import rdkit.RDLogger as rkl

        self.rdkit_logger = rkl.logger()
        print()
        #self.rdkit_logger.addHandler(self.fh)

        self.streamhandle_stdout = logging.StreamHandler(sys.__stdout__)
        self.streamhandle_stdout.setLevel(logging.DEBUG)
        self.streamhandle_stdout.setFormatter(self.formatter)
        self.logger.addHandler(self.streamhandle_stdout)



        self.streamhandle_stderr = logging.StreamHandler(sys.__stderr__)
        self.streamhandle_stderr.setLevel(logging.DEBUG)
        self.streamhandle_stderr.setFormatter(self.formatter)
        self.logger.addHandler(self.streamhandle_stderr)

        sys.stdout = StreamToLogger(self.logger,logging.DEBUG)
        # can't get this to work for stderr possibly because they can be the same
        #, but a battle for another day, 
```

**Comment by apayne97** [comment]
@hmacdope thanks for digging into this! 

Can you generate a simple example that demonstrates the problem? Maybe we can email OE Support and have them output to the python stdout instead? i.e. with something like https://pybind11.readthedocs.io/en/stable/advanced/pycpp/utilities.html (?)

**Comment by hmacdope** [comment]
Unlikely that they will want to rewrite their SWIG wrapper in PyBind11 (they are a pain to get working) but I may have found a solution. 

Through a combination of 
- https://docs.eyesopen.com/toolkits/python/oechemtk/OEPlatformConstants/OEPlatform.html#OEPlatform::OEPlatform::oeerr
- https://docs.eyesopen.com/toolkits/python/oechemtk/OESystemConstants/OEThrow.html#OESystem::OEThrow
- https://docs.eyesopen.com/toolkits/python/oechemtk/OESystemClasses/OEErrorHandler.html#oesystem-oeerrorhandler

I'll let you know how I go

**Comment by apayne97** [comment]
Thanks for this! I've implemented a version of what you've suggested in #171 into #104 for the prep_mers_pdbs.py script and it works well! https://github.com/choderalab/asapdiscovery/pull/104/commits/c1860add209f340188ddb8970fb004c0a653e683

Your logger implementation is probably still useful though, we should probably go through all the scripts and replace print with logger where useful. For this issue, I think I'll put in the description which method is successful and close the issue.
